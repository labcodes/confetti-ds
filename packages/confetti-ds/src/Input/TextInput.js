import React from "react";
import PropTypes from "prop-types";

import AbstractTextInput from "./AbstractTextInput";

export default class TextInput extends React.Component {
  static propTypes = {
    /** Text that will serve as unique identifier. It's also an important accessibility tool. */
    id: PropTypes.string.isRequired,
    /** The Input's text label. */
    label: PropTypes.string.isRequired,
    /** Disables the text input. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled: PropTypes.bool,
    /** Disables the text input. Won't be read by screen readers. */
    disabled: PropTypes.bool,
    /** Defines a default value for the Input initialization. */
    defaultValue: PropTypes.string,
    /** Text that will be rendered inside the Input field. */
    value: PropTypes.string,
    /** Defines which symbol to show. */
    icon: PropTypes.string,
    /** Defines the color of the displayed icon. */
    iconColor: PropTypes.string,
    /** Defines if the Input is required. */
    required: PropTypes.bool,
    /** Text that will be displayed as a help message below the input. */
    helpMessage: PropTypes.string,
    /** Text that will be displayed at the left portion of the Input. */
    prefix: PropTypes.string,
    /** Text that will be displayed at the right portion of the Input. */
    suffix: PropTypes.string,
    /** Defines if the Input is valid. */
    isValid: PropTypes.bool,
    /** Custom error message displayed below the Input when the value is not valid. */
    customErrorMsg: PropTypes.string,
    /** Callback action to be executed when the Input default value changes. */
    onChange: PropTypes.func,
    /** Callback action to be executed when the Input's Icon is clicked.  */
    onIconClick: PropTypes.func,
  };

  static defaultProps = {
    disabled: false,
    ariaDisabled: false,
    defaultValue: undefined,
    value: undefined,
    icon: undefined,
    iconColor: "mineral-70",
    required: false,
    helpMessage: undefined,
    prefix: undefined,
    suffix: undefined,
    isValid: undefined,
    customErrorMsg: undefined,
    onChange: () => {},
    onIconClick: () => {},
  };

  render() {
    const {
      id,
      label,
      defaultValue,
      value,
      icon,
      iconColor,
      required,
      helpMessage,
      prefix,
      suffix,
      isValid,
      customErrorMsg,
      onChange,
      onIconClick,
      disabled,
      ariaDisabled,
    } = this.props;

    return (
      <AbstractTextInput
        type="text"
        id={id}
        label={label}
        defaultValue={defaultValue}
        value={value}
        icon={icon}
        iconColor={iconColor}
        required={required}
        helpMessage={helpMessage}
        prefix={prefix}
        suffix={suffix}
        isValid={isValid}
        customErrorMsg={customErrorMsg}
        onChange={onChange}
        onIconClick={onIconClick}
        disabled={(!ariaDisabled && disabled) || undefined}
        ariaDisabled={ariaDisabled || undefined}
      />
    );
  }
}
