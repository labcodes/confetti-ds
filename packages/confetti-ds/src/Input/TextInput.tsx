import React, {SyntheticEvent} from "react";
import PropTypes from "prop-types";

import AbstractTextInput from "./AbstractTextInput";
import { IconTypes, ICON_TYPES, IconColorTypes, ICON_COLORS } from "../constants";

interface TextInputProps {
    /** Text that will serve as unique identifier. It's also an important accessibility tool. */
    id: string;
    /** The Input's text label. */
    label: string;
    /** Disables the text input. Will be read by screen readers. When true, will override `disabled`. */
    ariaDisabled?: boolean;
    /** Disables the text input. Won't be read by screen readers. */
    disabled?: boolean;
    /** Defines a default value for the Input initialization. */
    defaultValue?: string;
    /** Text that will be rendered inside the Input field. */
    value?: string;
    /** Defines which symbol to show. */
    icon?: IconTypes;
    /** Defines the color of the displayed icon. */
    iconColor?: IconColorTypes;
    /** Defines if the Input is required. */
    required?: boolean;
    /** Text that will be displayed as a help message below the input. */
    helpMessage?: string;
    /** Text that will be displayed at the left portion of the Input. */
    prefix?: string;
    /** Text that will be displayed at the right portion of the Input. */
    suffix?: string;
    /** Defines if the Input is valid. */
    isValid?: boolean;
    /** Custom error message displayed below the Input when the value is not valid. */
    customErrorMsg?: string;
    /** Callback action to be executed when the Input default value changes. */
    onChange?: (event?: SyntheticEvent) => any;
    /** Callback action to be executed when the Input's Icon is clicked.  */
    onIconClick?: (event?: SyntheticEvent) => any;
}

export default function TextInput({
  disabled = false,
  ariaDisabled = false,
  defaultValue,
  value,
  icon,
  iconColor = "mineral-70",
  required = false,
  helpMessage,
  prefix,
  suffix,
  isValid,
  customErrorMsg,
  onChange= () => {},
  onIconClick= () => {},
  id,
  label
}: TextInputProps) {
  return (
    <AbstractTextInput
      type="text"
      id={id}
      label={label}
      defaultValue={defaultValue}
      value={value}
      icon={icon}
      iconColor={iconColor}
      required={required}
      helpMessage={helpMessage}
      prefix={prefix}
      suffix={suffix}
      isValid={isValid}
      customErrorMsg={customErrorMsg}
      onChange={onChange}
      onIconClick={onIconClick}
      disabled={(!ariaDisabled && disabled) || undefined}
      ariaDisabled={ariaDisabled || undefined}
    />
  );
}

