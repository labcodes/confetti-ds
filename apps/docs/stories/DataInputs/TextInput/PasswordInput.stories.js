import React from "react";

import { PasswordInput as Component } from "../../../../../packages/confetti-ds/src/Input";

export default {
  title: "Data Inputs/Text Input/Password Input",
  component: Component,
};

export const PasswordInput = (args) => <Component {...args} />;
PasswordInput.args = {
  id: "demo-input",
  label: "Demo label",
  onChange: (event) => console.log(event),
};
