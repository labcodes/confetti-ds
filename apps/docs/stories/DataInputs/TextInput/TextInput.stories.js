import React from "react";

import { ICON_TYPES, ICON_COLORS } from "../../../confetti-ds/src/constants";
import { TextInput as Component } from "../../../confetti-ds/src/Input";

export default {
  title: "Data Inputs/Text Input/Text Input",
  component: Component,
  argTypes: {
    icon: {
      control: { type: "select", options: ICON_TYPES },
    },
    iconColor: {
      control: { type: "select", options: ICON_COLORS },
    }
  },
};

export const TextInput = (args) => <Component {...args} />;
TextInput.args = {
  id: "demo-input",
  label: "Demo label",
  onChange: (event) => console.log(event),
};
