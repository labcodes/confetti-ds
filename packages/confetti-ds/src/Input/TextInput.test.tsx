/* eslint-disable no-console */
import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";

import TextInput from "./TextInput";

describe("TextInput", () => {
  const originalWarn = console.warn;
  afterEach(() => {
    console.warn = originalWarn;
  });

  it("renders with base props", async () => {
    const renderedComponent = renderer
      .create(<TextInput id="testInput" label="Test Input" />)
      .toJSON();
    expect(renderedComponent).toMatchSnapshot();
  });

  it("renders as type=text", async () => {
    const component = mount(<TextInput id="testInput" label="Test Input" />);
    expect(component.find("input[type='text']")).toHaveLength(1);
  });

  it("raises console.warn and sets state with value when passing value and defaultValue by props at the same time", async () => {
    console.warn = jest.fn();

    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        defaultValue="default value"
        value="test value"
      />
    );

    expect(console.warn).toBeCalled();
    const inputElement = component.find("input");
    expect(inputElement.render().attr("value")).toBe("test value");
  });

  it("sets state with value if it is passed by props", async () => {
    const component = mount(
      <TextInput id="testInput" label="Test Input" value="test value" />
    );

    const inputElement = component.find("input");
    expect(inputElement.render().attr("value")).toBe("test value");
  });

  it("sets state with defaultValue if it is passed by props and value is not passed by props", async () => {
    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        defaultValue="default value"
      />
    );
    const inputElement = component.find("input");
    expect(inputElement.render().attr("value")).toBe("default value");
  });

  it("sets state with isValid if it is passed by props", async () => {
    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        defaultValue="default value"
        isValid={false}
      />
    );
    expect(component.find(".lab-input--invalid")).toHaveLength(1);
  });

  it("sets localIsValid to true if value is valid and change it if value validity changes", async () => {
    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        value="testvalue@g.com"
        required
      />
    );
    expect(component.find(".lab-input--invalid")).toHaveLength(0);

    component.setProps({ value: "" });
    component.update();

    expect(component.find(".lab-input--invalid")).toHaveLength(1);

    component.setProps({ value: "valid@value.com" });
    component.update();

    expect(component.find(".lab-input--invalid")).toHaveLength(0);
  });

  it("shows help text", async () => {
    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        value="testvalue@g.com"
        helpMessage="help message"
        required
      />
    );

    expect(component.find(".lab-input__message--required").text()).toBe(
      "help message"
    );

    component.setProps({ value: "" });
    component.update();

    expect(component.find(".lab-input__message--error").text()).toBe(
      "help message"
    );
  });

  it("shows custom error message when input is invalid", async () => {
    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        value="test"
        customErrorMsg="error message"
        required
      />
    );

    expect(component.find(".lab-input__message--error")).toHaveLength(0);

    component.setProps({ value: "" });
    component.update();

    expect(component.find(".lab-input__message--error").text()).toBe(
      "error message"
    );
  });

  it("changes localValue state when input changes", async () => {
    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        value="truthy value"
        onChange={() => {}}
      />
    );

    const inputElement = component.find("input");
    expect(inputElement.render().attr("value")).toBe("truthy value");

    component
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "My new value",
          validity: { valid: true },
          setCustomValidity: jest.fn(),
        },
      });

    expect(inputElement.render().attr("value")).toBe("My new value");
  });

  it("sets input as invalid if it is required and is not filled, even if isValid is passed by props as true", async () => {
    const component = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        isValid
        required
      />
    );

    expect(component.find(".lab-input--invalid")).toHaveLength(0);

    component
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "My new value",
          validity: { valid: true },
          setCustomValidity: jest.fn(),
        },
      });

    expect(component.find(".lab-input--invalid")).toHaveLength(0);

    component
      .find("input")
      .at(0)
      .simulate("change", {
        target: {
          value: "",
          validity: { valid: false },
          setCustomValidity: jest.fn(),
        },
      });

    expect(component.find(".lab-input--invalid")).toHaveLength(1);
  });

  it("renders prefix when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(<TextInput id="testInput" label="Test Input" prefix="R$" />)
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        prefix="R$"
      />
    );

    expect(mountedComponent.find("span.lab-input__prefix")).toHaveLength(2);
  });

  it("renders suffix when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(<TextInput id="testInput" label="Test Input" suffix=".com" />)
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        suffix=".com"
      />
    );

    expect(mountedComponent.find("div.lab-input__suffix")).toHaveLength(1);
  });

  it("renders icon when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(<TextInput id="testInput" label="Test Input" icon="star" />)
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <TextInput
        id="testInput"
        label="Test Input"
        icon="star"
      />
    );

    expect(mountedComponent.find("span.lab-icon")).toHaveLength(1);
  });

  it("renders required icon when it is passed by props", async () => {
    const renderedComponent = renderer
      .create(<TextInput id="testInput" label="Test Input" icon="star" />)
      .toJSON();

    expect(renderedComponent).toMatchSnapshot();

    const mountedComponent = mount(
      <TextInput id="testInput" label="Test Input" required />
    );

    expect(mountedComponent.find("span.lab-input__required-icon")).toHaveLength(
      1
    );
  });

  it("doesn't trigger onChange if ariaDisabled", async () => {
    const mockOnChange = jest.fn();
    const mountedComponent = mount(
      <TextInput
        ariaDisabled
        id="testInput"
        label="Test Input"
        defaultValue="default"
        onChange={mockOnChange}
      />
    );

    expect(mountedComponent.find("input").render().attr("value")).toBe(
      "default"
    );
    expect(mockOnChange).not.toBeCalled();

    mountedComponent.find("input").at(0).simulate("change", { test: "event" });

    expect(mountedComponent.find("input").render().attr("value")).toBe(
      "default"
    );
    expect(mockOnChange).not.toBeCalled();
  });
});
