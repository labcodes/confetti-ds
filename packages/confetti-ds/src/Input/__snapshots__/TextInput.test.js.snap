// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextInput renders icon when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  <button
    className="lab-input__icon"
    onClick={[Function]}
    type="button"
  >
    <span
      className="lab-icon lab-icon--star lab-icon--mineral-70"
    />
  </button>
  
</div>
`;

exports[`TextInput renders prefix when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field lab-input__field--prefixed "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  <span
    className="lab-input__prefix"
  >
    R$
  </span>
  
  <div
    className="lab-input__label-wrapper"
  >
    <span
      className="lab-input__prefix"
    >
      R$
    </span>
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  
</div>
`;

exports[`TextInput renders required icon when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  <button
    className="lab-input__icon"
    onClick={[Function]}
    type="button"
  >
    <span
      className="lab-icon lab-icon--star lab-icon--mineral-70"
    />
  </button>
  
</div>
`;

exports[`TextInput renders suffix when it is passed by props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field lab-input__field--suffixed "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  <div
    className="lab-input__suffix"
  >
    .com
  </div>
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  
</div>
`;

exports[`TextInput renders with base props 1`] = `
<div
  className="lab-input "
>
  <input
    autoComplete="off"
    className="lab-input__field "
    id="testInput"
    name="testInput"
    onChange={[Function]}
    placeholder=" "
    type="text"
    value=""
  />
  <div
    className="lab-input__borders"
  />
  
  
  <div
    className="lab-input__label-wrapper"
  >
    
    <label
      className="lab-input__label"
      htmlFor="testInput"
    >
      Test Input
    </label>
  </div>
  
</div>
`;
